utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala

```diff
@@ -0,0 +1,181 @@
+ */
+object Forest extends Serializable {
+
+  /**
+   * Sorts the RDD and collects it to build the tree.
+   *
+   * @param rdd RDD to build a forest from.
+   * @return The forest built from this RDD.
+   */
+  def apply[K <: Interval[K]: ClassTag, T: ClassTag](rdd: RDD[(K, T)]): Forest[K, T] = {
+    val sortedArray =
+      rdd.sortByKey()
+        .collect
+
+    new Forest(sortedArray)
+  }
+}
+
+/**
+ * Originally, a forest was a collection of trees.
+ * Alas, we have no trees anymore.
+ * I blame global warming.
+ *
+ * @param array An array of values for the left side of the join. We require
+ *   this array to be sorted.
+ */
+case class Forest[K <: Interval[K], T: ClassTag](array: Array[(K, T)]) {
+
+  val length = array.length
+  val midpoint = pow2ceil()
+
+  @tailrec private def pow2ceil(i: Int = 1): Int = {
+    if (2 * i >= length) {
+      i
+    } else {
+      pow2ceil(2 * i)
+    }
+  }
+
+  @tailrec private def binarySearch(rr: K,
+                                    idx: Int = 0,
+                                    step: Int = midpoint): Option[Int] = {
+    if (array.length == 0) {
+      None
+    } else if (rr.overlaps(array(idx)._1)) {
+      Some(idx)
+    } else if (step == 0) {
+      None
+    } else {
+      val stepIdx = idx + step
+      val nextIdx = if (stepIdx >= length ||
+        (!rr.overlaps(array(stepIdx)._1) &&
+          rr.compareTo(array(stepIdx)._1) < 0)) {
+        idx
+      } else {
+        stepIdx
+      }
+      binarySearch(rr, nextIdx, step / 2)
+    }
+  }
+
+  @tailrec private def expand(rr: K,
+                              idx: Int,
+                              step: Int,
+                              list: List[(K, T)] = List.empty): List[(K, T)] = {
+    if (idx < 0 ||
+      idx >= length ||
+      !rr.overlaps(array(idx)._1)) {
+      list
+    } else {
+      expand(rr, idx + step, step, array(idx) :: list)
+    }
+  }
+
+  /**
+   * Insert Iterator of (K,V) items into existing Forest
+   * @param kvs (K,V) tuples to insert into Forest
+   * @return new Forest with inserted values
+   */
+  def insert(kvs: Iterator[(K, T)]): Forest[K, T] = {
+    Forest((array ++ kvs).sortBy(_._1))
+  }
+
+  /**
+   * Filters items in Forest based on predicate on (K,V) tuples
+   * @param pred predicate to filter elements by
+   * @return new forest with filtered elements
+   */
+  def filter(pred: ((K, T)) => Boolean): Forest[K, T] = {
+    Forest(array.filter(r => pred(r._1, r._2)))
+  }
+
+  /**
+   * Maps values from T to T2
+   * @param f Function mapping T to T2
+   * @tparam T2 new type to map values to
+   * @return new Forest with mapped values
+   */
+  def mapValues[T2: ClassTag](f: T => T2): Forest[K, T2] = {
+    Forest(array.map(r => (r._1, f(r._2))))
+  }
+
+  /**
+   * Filters elements in array by overlapping Interval
+   * @param rr Interval to filter by
+   * @return Iterable of elements filtered by Interval rr
+   */
+  def get(rr: K): Iterable[(K, T)] = {
```

Comment:
@fnothaft I changed the return type from Iterable[T] to  Iterable[(K, T)] for purposes in IntervalRDD. Will this be a problem for the Tree Join? I was going to add in both but the API got a little messy.